File: Design/208_ImplementTrie(PrefixTree).py
class TrieNode(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.children = [None]*26
        self.isword = False
        self.char = None

class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        wd = self.root
        for c in word:
            if wd.children[ord(c)-ord("a")] is None:
                wd.children[ord(c)-ord("a")] = TrieNode()
                wd.children[ord(c)-ord("a")].char = c
            wd = wd.children[ord(c)-ord("a")]
        wd.isword = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        wd = self.root
        for c in word:
            cld = wd.children[ord(c)-ord("a")]
            if cld is None:
                return False
            wd = cld
        return wd.isword

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        wd = self.root
        for c in prefix:
            cld = wd.children[ord(c)-ord("a")]
            if cld is None:
                return False
            wd = cld
        return True

# Your Trie object will be instantiated and called as such:
# trie = Trie()
# trie.insert("somestring")
# trie.search("key")

File: Design/208_ImplementTrie(PrefixTree).java
class TrieNode {
    // Initialize your data structure here.
	boolean isWord;
	TrieNode[] children;
    public TrieNode() {
        this.children = new TrieNode[26];
    }
}

public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    public void insert(String word) {
        TrieNode curr = this.root;
		for(char c : word.toCharArray()) {
			TrieNode tn = curr.children[c-'a'];
			if(tn == null)
				tn = new TrieNode();
			curr = tn;
		}
		curr.isWord = true;
    }

    // Returns if the word is in the trie.
    public boolean search(String word) {
        TrieNode curr = this.root;
		for(char c : word.toCharArray()) {
			TrieNode tn = curr.children[c-'a'];
			if(tn == null)
				return false;
			curr = tn;
		class TrieNode {
    // Initialize your data structure here.
	boolean isWord;
	TrieNode[] children;
    public TrieNode() {
        this.children = new TrieNode[26];
    }
}

public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    // Inserts a word into the trie.
    public void insert(String word) {
        TrieNode curr = this.root;
		for(char c : word.toCharArray()) {
			if(curr.children[c-'a'] == null)
				curr.children[c-'a'] = new TrieNode();
			curr = curr.children[c-'a'];
		}
		curr.isWord = true;
    }

    // Returns if the word is in the trie.
    public boolean search(String word) {
        TrieNode curr = this.root;
		for(char c : word.toCharArray()) {
			TrieNode tn = curr.children[c-'a'];
			if(tn == null)
				return false;
			curr = tn;
		}
		return curr.isWord;
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    public boolean startsWith(String prefix) {
        TrieNode curr = this.root;
		for(char c : prefix.toCharArray()) {
			TrieNode tn = curr.children[c-'a'];
			if(tn == null)
				return false;
			curr = tn;
		}
		return true;
    }
}

// Your Trie object will be instantiated and called as such:
// Trie trie = new Trie();
// trie.insert("somestring");
// trie.search("key");
		return curr.isWord;
    }

    // Returns if there is any word in the trie
    // that starts with the given prefix.
    public boolean startsWith(String prefix) {
        TrieNode curr = this.root;
		for(char c : word.toCharArray()) {
			TrieNode tn = curr.children[c-'a'];
			if(tn == null)
				return false;
			curr = tn;
		}
		return true;
    }
}

// Your Trie object will be instantiated and called as such:
// Trie trie = new Trie();
// trie.insert("somestring");
// trie.search("key");